### PR TITLE
[cherry-pick 2.0rc1][inplace] Add ShareHolderWith for class Variable and SharePlaceholderWith in VarBase.detach() to share the same Tensor/SelectedRows (#29267)

### PR DESCRIPTION
**Pull request by liym27**
### PR types

Others
### PR changes

Others
### Describe

cherry piick https://github.com/PaddlePaddle/Paddle/pull/29267 to release/2.0-rc1

### CODE REVIEW

**Review by zhhsplendid — APPROVED** [review]
LGTM
